sql/analyzer/resolve_unions.go

```diff
@@ -22,66 +22,83 @@ import (
 	if n.Resolved() {
-		return n, nil
+		return n, transform.SameTree, nil
 	}
 	// Procedures explicitly handle unions
 	if _, ok := n.(*plan.CreateProcedure); ok {
-		return n, nil
+		return n, transform.SameTree, nil
 	}
 
-	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
-		switch n := n.(type) {
+	return transform.Node(n, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch n := node.(type) {
 		case *plan.Union:
 			subqueryCtx, cancelFunc := ctx.NewSubContext()
 			defer cancelFunc()
 
-			left, err := a.analyzeThroughBatch(subqueryCtx, n.Left(), scope, "default-rules")
+			left, sameL, err := a.analyzeThroughBatch(subqueryCtx, n.Left(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			right, err := a.analyzeThroughBatch(subqueryCtx, n.Right(), scope, "default-rules")
+			right, sameR, err := a.analyzeThroughBatch(subqueryCtx, n.Right(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			return n.WithChildren(StripPassthroughNodes(left), StripPassthroughNodes(right))
+			if sameL && sameR {
+				return node, transform.SameTree, nil
+			}
+			node, err = n.WithChildren(StripPassthroughNodes(left), StripPassthroughNodes(right))
+			if err != nil {
+				return nil, transform.SameTree, err
+			}
+			return node, transform.NewTree, nil
+
 		default:
-			return n, nil
+			return n, transform.SameTree, nil
 		}
 	})
 }
 
-func finalizeUnions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, error) {
+func finalizeUnions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope) (sql.Node, transform.TreeIdentity, error) {
 	// Procedures explicitly handle unions
 	if _, ok := n.(*plan.CreateProcedure); ok {
-		return n, nil
+		return n, transform.SameTree, nil
 	}
 
-	return plan.TransformUp(n, func(n sql.Node) (sql.Node, error) {
-		switch n := n.(type) {
+	return transform.Node(n, func(node sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		switch n := node.(type) {
 		case *plan.Union:
 			subqueryCtx, cancelFunc := ctx.NewSubContext()
 			defer cancelFunc()
 
-			left, err := a.analyzeStartingAtBatch(subqueryCtx, n.Left(), scope, "default-rules")
+			// TODO we could detect tree modifications here, skip rebuilding
+			left, sameL, err := a.analyzeStartingAtBatch(subqueryCtx, n.Left(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			right, err := a.analyzeStartingAtBatch(subqueryCtx, n.Right(), scope, "default-rules")
+			right, sameR, err := a.analyzeStartingAtBatch(subqueryCtx, n.Right(), scope, "default-rules")
 			if err != nil {
-				return nil, err
+				return nil, transform.SameTree, err
 			}
 
-			return n.WithChildren(StripPassthroughNodes(left), StripPassthroughNodes(right))
+			if sameL && sameR {
```

Comment:
Same comment, probably just want to ignore the `same` result from `Analyze` until we get rid of NodesEqual